### PR TITLE
[Caliban] Use apache's async http client

### PR DESCRIPTION
**Pull request by kyri-petrou**
Offers slightly improved throughput than the JDK11's stock HTTP client. Ideally we'd like to use zio-http's (Netty) client but unfortunately it doesn't seem to work with the proxy 😕

### CODE REVIEW

**Comment by tusharmath** [comment]
I have realised sort of flaw in the current benchmarking architecture. I think both Tailcall and Caliban are reaching their maximum performance and are being limited by the upstream nginx.

This ultimately creates quite a noisy environment for benchmarking individual servers. I think not depending on an external service for data, essentially removing all IOs, would be a much better benchmarking test. That will truly test the performance of the graphQL servers.

**Comment by tusharmath** [comment]
I wonder how tailcall would perform if we increased the connection pool size and TTL like it's done on caliban.

**Comment by tusharmath** [comment]
Actually the benchmark results below, showcase a more realistic difference in performance — 

https://github.com/graphql-crystal/benchmarks/actions/runs/6922987306/job/18830350044#step:8:886 

Tailcall's first version was written in Scala using ZIO Http and Caliban, we eventually migrated to Rust because we achived better performance using Rust. In your benchmarks, you end up beating Tailcall sometimes, which truly flips me out and I get into this worm hole of - could the benchmarks be wrong?

**Comment by kyri-petrou** [comment]
@tusharmath there are a couple of things to unpack here:

> Actually the benchmark results below, showcase a more realistic difference in performance

The query benchmarked here requires extremely trivial graphql execution since the field returns a simple string. This means that the benchmark is largely biased by the HTTP server's performance. The Rust and Go HTTP servers are definitely faster than ZIO HTTP which explains this difference.

This was also the reason that the benchmarks when the proxy was not working properly showed Tailcall & Glgen being much faster than Caliban - simply because the top-level field returned an error and therefore the benchmark became biased towards the HTTP server's performance

> Tailcall's first version was written in Scala using ZIO Http and Caliban, we eventually migrated to Rust because we achived better performance using Rust.

Starting with Caliban 2.2 we performed extensive optimizations in Caliban's validator & executor. Prior to Caliban 2.3, Caliban's executor was roughly x2 faster than Sangria's, but now it's more than x10 faster according to our benchmarks. In addition, in version 2.4.3, we removed the intermediate tapir layer in the `caliban-quick` module and reimplemented the HTTP layer using zio-http, which significantly improved performance.

In short, the current version of Caliban is significantly far more performant than prior versions

**Comment by tusharmath** [comment]
> The query benchmarked here requires extremely trivial graphql execution since the field returns a simple string. This means that the benchmark is largely biased by the HTTP server's performance. The Rust and Go HTTP servers are definitely faster than ZIO HTTP which explains this difference.

I think http-server's performance is relevant when it comes to building a GraphQL service at the same time I also think that it's not enough. We definitely need a suite exhaustive tests to truly compare graphQL implementations. Broadly speaking there are three components to a GraphQL service:

1. Ingress: That's what crystal benchmarks are testing.
2. Egress: That's what our benchmarks test.
3. GraphQL Execution: This includes, query parsing, executing resolvers and creating a GraphQL Response. A heavy recursive query/mutation would be a great test to try it out.

Let me know what you think, and if I have missed something. 

PS: Benchmarking can quickly become a rabbit hole, so most people give up after optimizing for a while. I hope we can continue working be really transparent about the results. With ZIO Http we set a really high bar for ourselves, which is really great :)
